students/tri_nguyen/lesson04/mailroom.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+
+# ------------------------------------------- #
+# Student:   Tri Nguyen
+# Instructor: Natasha Aleksandrova
+# Date:  25-Feb-2018
+# ------------------------------------------- #
+
+# Import stuff
+from operator import itemgetter
+
+# Declare variables and data structure
+
+donor_dict = [
+    {'name': 'William Gates, III', 'donation': 653784.49, 'count': 2, 'average': 326892.24},
+    {'name': 'Mark Zuckerberg', 'donation': 16396.10, 'count': 3, 'average': 5465.37},
+    {'name': 'Jeff Bezos', 'donation': 877.33, 'count': 1, 'average': 877.33},
+    {'name': 'Paul Allen', 'donation': 708.42, 'count': 3, 'average': 236.14},
+    {'name': 'Tri Nguyen', 'donation': 100.00, 'count': 1, 'average': 100.00}
+]
```

Comment:
although this data structure can work, it is not ideal.

in this particular case `name` is our unique key for donors. consider that you need to check if donor exists to determine if you should add donations to existing one or if you need to add new donor, this is what you do:
```
for donor in donor_list:
    if donor['name'] == name_entered:
        # append to existing donor
```
1. it's way more complex to manage donors 2. it is expensive (imagine you have a large list of donors)

now if you actually used a dictionary:
```
donor_dict = {"Paul Allen": [10, 100, 120], "Jeff Bezos": [3, 44, 45]}
if donor_dict.get(name_entered):
    donor_dict[name_entered].append(new_donation)
else:
    donor_dict[name_entered] = [new_donation]
```
1. simple, no more looping looking for something in the list 2. it's *cheap*, O(1) time which means that, no matter how much data, it will execute in constant time (that is not the case if a list look up).

Reply:
Thanks, Natasha. Your instructions and explanations here are very helpful. I kind of knew that there would be a better way to organize. I will make changes.

Reply:
👍